src/xenia/kernel/xam/xam_user.cc

```diff
@@ -552,23 +552,62 @@ dword_result_t XamShowSigninUI(dword_t unk, dword_t unk_mask) {
 }
 DECLARE_XAM_EXPORT1(XamShowSigninUI, kUserProfiles, kStub);
 
+#pragma pack(push, 1)
+struct X_XACHIEVEMENT_DETAILS {
+  xe::be<uint32_t> id;
+  xe::be<uint32_t> label_ptr;
+  xe::be<uint32_t> description_ptr;
+  xe::be<uint32_t> unachieved_ptr;
+  xe::be<uint32_t> image_id;
+  xe::be<uint32_t> gamerscore;
+  xe::be<uint64_t> unlock_time;
+  xe::be<uint32_t> flags;
+};
+static_assert_size(X_XACHIEVEMENT_DETAILS, 36);
+#pragma pack(pop)
+
 dword_result_t XamUserCreateAchievementEnumerator(dword_t title_id,
                                                   dword_t user_index,
                                                   dword_t xuid, dword_t flags,
                                                   dword_t offset, dword_t count,
                                                   lpdword_t buffer_size_ptr,
                                                   lpdword_t handle_ptr) {
+  if (!count || !buffer_size_ptr || !handle_ptr) {
+    return X_ERROR_INVALID_PARAMETER;
+  }
+
+  if (user_index >= 4) {
+    return X_ERROR_INVALID_PARAMETER;
+  }
+
+  // Providing flag 0x20 only gives you basic info about achievements
+  uint32_t entry_size = 500;
+  if (flags & 0x20) {
+    entry_size = sizeof(X_XACHIEVEMENT_DETAILS);
+  }
+
   if (buffer_size_ptr) {
-    *buffer_size_ptr = 500 * count;
+    *buffer_size_ptr = entry_size * count;
   }
 
   auto e = object_ref<XStaticEnumerator>(
-      new XStaticEnumerator(kernel_state(), count, 500));
+      new XStaticEnumerator(kernel_state(), count, entry_size));
   auto result = e->Initialize(user_index, 0xFB, 0xB000A, 0xB000B, 0);
   if (XFAILED(result)) {
     return result;
   }
 
+  // Titles can provide incorrect achievements count
+  // For example "Way of the Dogg" provides 0x80 of them
```

Comment:
What do you mean by "incorrect"? It should match the allocated memory size, not the number of achievements the game has.

Reply:
Fair point. It's not amount of achievements that game have, but amount of entries for which we should allocate memory.

So, even it game have let's say 10 achievements, it can request space for 20 items.